build.rs

```diff
@@ -351,7 +347,105 @@ fn build_msvc(target_triple: &[&str], disable_opt: bool, num_jobs: &str,
     let mut test_args = args.clone();
     test_args.push(String::from("crypto/libring-test.Windows.vcxproj"));
     run_command_with_args(&msbuild, &test_args);
-    Ok(())
+    println!("cargo:rustc-link-search=native={}", lib_path.to_str().unwrap());
+    println!("cargo:rustc-link-lib=static={}-core", LIB_NAME);
+
+    // XXX: Ideally, this would only happen for `cargo test`, but we don't know
+    // how to do that yet.
+    println!("cargo:rustc-link-lib=static={}-test", LIB_NAME);
+}
+
+fn build_unix(target_info: &TargetInfo, out_dir: PathBuf) {
+    let mut lib_target = out_dir.clone();
+    lib_target.push("libring-core.a");
+    let lib_target = lib_target.as_path();
+
+    let mut test_target = out_dir.clone();
+    test_target.push("libring-test.a");
+    let test_target = test_target.as_path();
+
+    let lib_header_change = RING_HEADERS.iter()
+        .map(Path::new)
+        .any(|p| need_run(&p, lib_target));
+    let test_header_change = RING_TEST_HEADERS.iter()
+        .map(Path::new)
+        .any(|p| need_run(&p, test_target)) ||
+                             lib_header_change;
+
+    let mut additional = Vec::new();
+    for additional_src in target_info.get_srcs() {
+        for src in additional_src.iter() {
+            additional.push(make_asm(src, out_dir.clone(), &target_info));
+        }
+    }
+
+    let objs = RING_SRC.iter()
+        .map(|a| String::from(*a))
+        .chain(additional.into_iter())
+        .map(|f| compile(&f, &target_info, out_dir.clone(), lib_header_change))
+        .collect::<Vec<_>>();
+
+
+    if objs.iter().map(|f| Path::new(f)).any(|p| need_run(&p, lib_target)) {
+
+        let mut c = gcc::Config::new();
+        for o in objs {
+            let _ = c.object(o);
+        }
+        c.compile("libring-core.a");
+    }
+
```

Comment:
Please factor out the duplicate logic between the code in this function above this line and the code below this line.